dbms/src/Storages/LiveView/StorageLiveView.cpp

```diff
@@ -453,21 +458,28 @@ void StorageLiveView::shutdown()
     if (!shutdown_called.compare_exchange_strong(expected, true))
         return;
 
-    if (no_users_thread.joinable())
     {
+        std::lock_guard no_users_thread_lock(no_users_thread_mutex);
+        if (no_users_thread.joinable())
         {
-            std::lock_guard lock(no_users_thread_mutex);
-            no_users_thread_wakeup = true;
-            no_users_thread_condition.notify_one();
+            {
+                std::lock_guard lock(no_users_thread_wakeup_mutex);
+                no_users_thread_wakeup = true;
+                no_users_thread_condition.notify_one();
+            }
         }
     }
 }
 
 StorageLiveView::~StorageLiveView()
 {
     shutdown();
-    if (no_users_thread.joinable())
-        no_users_thread.detach();
+
+    {
+        std::lock_guard lock(no_users_thread_mutex);
+        if (no_users_thread.joinable())
+            no_users_thread.detach();
```

Comment:
The fix is totally incorrect:

https://clickhouse-test-reports.s3.yandex.net/14404/7ca155c417a4ac64f448c390cc5f5462540b204f/stress_test_(thread)/stderr.log

Calling `std::thread::detach` is just guaranteed error.

Reply:
I don't think it is totally incorrect but I fully agree that calling and using `std::thread::detach` is bad. So what would be a better way to asynchronously drop table? The use case here is that if `TIMEOUT` is specified during the creation of `LIVE VIEW` we want to drop it automatically if there are no users that are watching this view after some `N` number of seconds.

Reply:
Maybe just remove the TIMEOUT feature?

Reply:
So you think that there is nothing in ClickHouse where we could schedule a callback?

Reply:
Maybe we should do something like this? https://github.com/ClickHouse/ClickHouse/blob/83a89fe858b03290b3c70eb0a6eb0e9145e40018/src/Storages/RabbitMQ/StorageRabbitMQ.cpp#L105

Reply:
Yes, it can be Ok.

Reply:
Ok, perfect. I will try it out and create a PR.